### PR TITLE
Add private auth support for PackageRepository with secretgen-controller

### LINKED ISSUE #242: Feature: PackageRepository can automatically auth to private registries when secretgen-controller is in the cluster
### Problem Description
Right now, the process of creating a package repository that can auth to a private registry is an entirely manual one. The user needs to create a secret with the authorization details, and then correctly reference it within their repository yaml. The manual nature of this makes it prone to human error.

### Solution Description
kapp-controller is able to automatically handle the auth details of a package repository.

#### 🟢 Public repo

**Given** A repo bundle stored in a public registry
**When** A user adds a PackageRepository CR with the repo's url
**And** Successfully fetches the repository bundle

**Note:** For now, since kapp-controller can't determine if a registry url is public or private, this may require creating a placeholder secret anyways.

---
#### 🟢 Private Repo

**Given** A repo bundle stored in a private registry
**When** A user adds a PackageRepository CR with the repo's url
**Then** kapp-controller creates a placeholder secret for the repository fetch
**And** Eventually kapp-controller successfully fetches the repository bundle

---
#### 🟢 Private Repo with explicit auth

**Given** A repo bundle stored in a private registry
**When** A user adds a PackageRepository CR with the repo's url and explicit auth configuration
**Then** kapp-controller **does not** create a placeholder secret for the repository fetch
**And** Eventually kapp-controller successfully fetches the repository bundle

---
#### 🔴 Credentials are not exported

**Given** a package repository with a fetch stage pulling from a private registry
**And** No secrets are exported (secretgen-controller behavior)
**When** a user creates a PackageRepository CR to fetch the repo
**Then** kapp-controller will create a placeholder secret
**And** Will add an error message detailing the auth failure from the registry when fetching

---
#### 🔴 Credentials are exported but not valid

**Given** a package repository with a fetch stage pulling from a private registry
**And** The exported credentials don't work (secretgen-controller behavior)
**When** a user creates a PackageRepository CR to fetch the repo
**Then** kapp-controller will create a placeholder secret
**And** Will add an error message detailing the auth failure from the registry when fetching

### More Info
To do this, we plan on adding a feature to secretgen-controller (SGC) called placeholder secrets. Essentially, there will be a namespace with some exported secret, and when SGC detects that a placeholder secret matching some criteria has been created in a another namespace, it will fill in the details of the placeholder with the same content as the exported secret. With this feature in place, kapp-controller will just need to create the appropriate placeholder secret and add it to the auth details of a repo during reconciliation.

---
Vote on this request

This is an invitation to the community to vote on issues, to help us prioritize our backlog. Use the "smiley face" up to the right of this comment to vote.

👍 "I would like to see this addressed as soon as possible"
👎 "There are other more important things to focus on right now"

We are also happy to receive and review Pull Requests if you want to help working on this issue.

### PR DESCRIPTION
**Pull request by danielhelfand**
Closes #242

### CODE REVIEW

File: cmd/controller/pkg_repository_reconciler.go [hunk 1]
```diff
@@ -66,6 +75,68 @@ func (r *PkgRepositoryReconciler) Reconcile(ctx context.Context, request reconci
 	return crdApp.Reconcile(force)
 }
 
+// ReconcileFetchPlaceholderSecrets helps determine if a placeholder secret
+// needs to be created for the PackageRepository. This placeholder secret is
+// populated by secretgen-controller so PackageRepositories can authenticate
+// to private registries without needing to explicitly declare a secretRef.
+// If no secretRef is specified for the PackageRepository, the placeholder
+// is created and used by the PackageRepository.
+func (r *PkgRepositoryReconciler) ReconcileFetchPlaceholderSecrets(pkgr v1alpha12.PackageRepository, app *v1alpha1.App) error {
```
Comment (cppforlife) [inline cmd/controller/pkg_repository_reconciler.go]: not ideal that this is publicly exposed on the reconciler, but it's fine for now i think, especially that package repo is implemented via internal App. might be good to play around with some refactoring post release.

File: cmd/controller/pkg_repository_reconciler.go [hunk 2]
```diff
@@ -66,6 +75,68 @@ func (r *PkgRepositoryReconciler) Reconcile(ctx context.Context, request reconci
...
+			if err != nil {
+				return err
+			}
+			app.Spec.Fetch[i].Image.SecretRef = &v1alpha1.AppFetchLocalRef{secretName}
+		}
+	}
+	return nil
+}
+
+func (r *PkgRepositoryReconciler) createSecretForSecretgenController(pkgr v1alpha12.PackageRepository, i int) (string, error) {
```
Comment (cppforlife) [inline cmd/controller/pkg_repository_reconciler.go]: nit: what was the name we were using in the app package for this?

File: cmd/controller/pkg_repository_reconciler.go [hunk 3]
```diff
@@ -5,20 +5,24 @@ package controller
...
 import (
 	"context"
-
-	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/apis/kappctrl/v1alpha1"
-	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/reftracker"
+	"fmt"
 
 	"github.com/go-logr/logr"
+	"github.com/vmware-tanzu/carvel-kapp-controller/pkg/apis/kappctrl/v1alpha1"
+	v1alpha12 "github.com/vmware-tanzu/carvel-kapp-controller/pkg/apis/packaging/v1alpha1"
```
Comment (cppforlife) [inline cmd/controller/pkg_repository_reconciler.go]: probably better to name this pkgv1alpha1.

File: cmd/controller/pkg_repository_reconciler.go (hunk 1 above)
Comment (danielhelfand) [reply cmd/controller/pkg_repository_reconciler.go]: This is mainly done for testing purposes with the function. Alternatively, the tests could be under the controller package as opposed to controller_test.

File: cmd/controller/pkg_repository_reconciler.go (hunk 2 above)
Comment (danielhelfand) [reply cmd/controller/pkg_repository_reconciler.go]: Do you mean packageinstall package? It was also named `createSecretForSecretgenController`.

File: cmd/controller/pkg_repository_reconciler.go (hunk 3 above)
Comment (danielhelfand) [reply cmd/controller/pkg_repository_reconciler.go]: Added

File: cmd/controller/pkg_repository_reconciler.go (hunk 2 above)
Comment (cppforlife) [reply cmd/controller/pkg_repository_reconciler.go]: oh ok. i thought it was renamed to Placeholder something nm.

File: cmd/controller/pkg_repository_reconciler.go (hunk 1 above)
Comment (cppforlife) [reply cmd/controller/pkg_repository_reconciler.go]: yup. just noting. controller_test is good.

**Pull request merged**